File: leetcode/solutions/find-the-duplicate-number/sol.cpp
// Problem: https://leetcode.com/problems/find-the-duplicate-number/

// Tort/Hare algorithm
// See https://leetcode.com/problems/find-the-duplicate-number/solution/

class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        // Tort and hare start at `nums[0]`, not `0`
        int tort = nums[0], hare = nums[0];
        do {
            tort = nums[tort];
            hare = nums[nums[hare]];
        } while (tort != hare);

        tort = nums[0];
        while (tort != hare) {
            tort = nums[tort];
            hare = nums[hare];
        }

        return tort;
    }
};

File: leetcode/solutions/find-the-duplicate-number/sol.py
#%%
from typing import List


class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        """
        NOTE: XORing will not work, because even though only a single number
        is repeated, it can be repeated multiple times.
        [3, 3, 3, 3, 3] is a valid input

        Solution: Floyd's algorithm: https://www.youtube.com/watch?v=wjYnzkAhcNk
        """

        slow, fast = 0, 0 # hold array indices

        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]
            if fast == slow:
                break

        slow2 = 0
        while slow != slow2:
            slow = nums[slow]
            slow2 = nums[slow2]

        return slow2

Solution().findDuplicate([1,1,1,2])

